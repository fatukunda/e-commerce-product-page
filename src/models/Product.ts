export interface IVariant {
  id: number;
  product_id: number;
  title: string;
  price: string;
  sku: string;
  position: number;
  compare_at_price: string;
  fulfillment_service: string;
  inventory_management: string;
  option1: string;
  option2?: string;
  option3?: string;
  created_at: string;
  updated_at: string;
  taxable: boolean;
  barcode: string;
  grams: number;
  image_id: number;
  weight: number;
  weight_unit: string;
  tax_code?: string;
  requires_shipping: boolean;
}

export interface IOption {
  id: number;
  product_id: number;
  name: string;
  position: number;
  values: string[];
}

export interface IImage {
  id: number;
  product_id: number;
  position: number;
  created_at: string;
  updated_at: string;
  alt?: string|null;
  width: number;
  height: number;
  src: string;
  variant_ids: number[];
}

export interface IProduct {
  id: number;
  title: string;
  body_html: string;
  vendor: string;
  product_type: string;
  created_at: string;
  handle: string;
  url: string;
  updated_at: string;
  published_at: string;
  variants: Array<
    {
      id: number;
      product_id: number;
      title: string;
      price: string;
      sku: string;
      position: number;
      compare_at_price: string;
      fulfillment_service: string;
      inventory_management: string;
      option1: string;
      option2?: string|null;
      option3?: string|null;
      created_at: string;
      updated_at: string;
      taxable: boolean;
      barcode: string;
      grams: number;
      image_id: number;
      weight: number;
      weight_unit: string;
      tax_code?: string;
      requires_shipping: boolean;
    }>
  options: [
    {
      id: number;
      product_id: number;
      name: string;
      position: number;
      values: string[];
    }
  ];
  images: Array<
    {
      id: number;
      product_id: number;
      position: number;
      created_at: string;
      updated_at: string;
      alt?: string|null;
      width: number;
      height: number;
      src: string;
      variant_ids: number[];
    }>;
  image: {
    id: number;
    product_id: number;
    position: number;
    created_at: string;
    updated_at: string;
    alt?: string|null;
    width: number;
    height: number;
    src: string;
    variant_ids: number[];
  };
}
