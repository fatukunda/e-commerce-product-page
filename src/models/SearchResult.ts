export interface ISearchItem {
    price: number,
    image: string,
    vendor: string,
    title: string,
    id: string,
    link: string
}