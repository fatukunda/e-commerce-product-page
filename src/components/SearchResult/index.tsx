import { FunctionComponent, useState, useEffect } from "react";
import Dropdown, { IOption } from "../Dropdown";
import Loader from "../Loader";

import ItemCard from "../ItemCard";

import { useAppSelector } from "src/store/hooks";
import { ISearchItem } from "src/models/SearchResult";
import Menu from "../Menu";
import { sort } from "src/services/searchService";
import { useAppDispatch } from "src/store/hooks";
import { setProducts, setIsSearching } from "src/store/slices/searchSlice";

const SearchResult: FunctionComponent = (): JSX.Element => {
  const searchedItems: ISearchItem[] = useAppSelector(
    (state) => state.search.items
  );
  const dispatch = useAppDispatch();
  const isSearching: boolean = useAppSelector(
    (state) => state.search.isSearching
  );
  const searchOptions: IOption[] = [
    {
      id: 1,
      name: "Random",
    },
    {
      id: 2,
      name: "Ascending",
    },
    {
      id: 3,
      name: "Descending",
    },
    {
      id: 4,
      name: "Price Low to High",
    },
    {
      id: 5,
      name: "Price High to Low",
    },
  ];
  const [selectedOption, setSelectedOption] = useState({
    id: 1,
    name: "Random",
  });
  const [isMenuOpen, setIsMenuOpen] = useState(false);
  const selectedOptionHandler = (option: IOption) => {
    setSelectedOption(option);
    setIsMenuOpen(false);
  };
  const closeMenuHandler = () => {
    setIsMenuOpen(false);
  };
  const toggleOpenHandler = () => {
    setIsMenuOpen(!isMenuOpen);
  };

  useEffect(() => {
    const search = async () => {
      let sortOption = "";
      if (selectedOption.name === "Ascending") {
        sortOption = "asc";
      } else if (selectedOption.name === "Descending") {
        sortOption = "desc";
      } else if (selectedOption.name === "Price Low to High") {
        sortOption = "price-low-high";
      } else if (selectedOption.name === "Price High to Low") {
        sortOption = "price-high-low";
      } else {
        sortOption = "Random";
      }
      dispatch(setIsSearching(true));
      const res = await sort(sortOption);
      dispatch(setIsSearching(false));
      dispatch(setProducts(res));
    };
    search();
  }, [selectedOption, dispatch]);

  return (
    <Menu classList="h-2/3 left-1/2 z-10 w-2/3  -translate-x-1/2 rounded-lg bg-white shadow-2xl shadow-tertiary/70 md:right-0 md:top-20  lg:top-24 lg:right-0 ">
      <div className="flex border-secondary-dark-dark block border-b border-solid p-6 font-bold items-center justify-between">
        <span className="block font-bold ">Your search results</span>
        <div>
          <Dropdown
            options={searchOptions}
            selectedItem={selectedOption}
            change={selectedOptionHandler}
            close={closeMenuHandler}
            isOpen={isMenuOpen}
            toggleOpen={toggleOpenHandler}
          />
        </div>
      </div>

      {!isSearching ? (
        <div className="p-4 flex gap-2 flex-wrap justify-center">
          {searchedItems?.length ? (
            searchedItems.map((item) => <ItemCard key={item.id} item={item} />)
          ) : (
            <div className="p-4 mt-36">
              <span className="inline-block w-full text-center font-bold text-tertiary-light">
                Oh no! No result for that search
              </span>
            </div>
          )}
        </div>
      ) : (
        <Loader />
      )}
    </Menu>
  );
};

export default SearchResult;
