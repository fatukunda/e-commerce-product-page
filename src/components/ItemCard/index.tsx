import { ISearchItem } from "src/models/SearchResult";

interface IItemCardProps {
  item: ISearchItem;
}

const ItemCard = ({ item }: IItemCardProps): JSX.Element => {
  return (
    <div className="w-48 bg-white rounded-lg shadow-md dark:border-gray-700">
        <img
          className="p-8 rounded-t-lg"
          src={item.image}
          alt={item.title}
        />
      <div className="px-5 pb-5">
        <h5 className="text-xs text-tertiary-light">{item.title}</h5>

        <div className="flex justify-between items-center">
          <span className="text-l pt-2 font-bold text-gray-900">
            ${item.price}
          </span>
        </div>
      </div>
    </div>
  );
};

export default ItemCard;
