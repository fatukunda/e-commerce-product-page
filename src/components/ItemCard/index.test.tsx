import { render, cleanup, fireEvent } from "@testing-library/react";
import renderer from "react-test-renderer";
import { Provider as ReduxProvider } from "react-redux";
import ItemCard from "./";
import mockStore from "../mocks/store";
import { ISearchItem } from "src/models/SearchResult";

const item: ISearchItem = {
  price: 20.22,
  image: "https://test-image",
  vendor: "test vendor",
  title: "Test product title",
  id: "1",
  link: "https://test-image-link",
};

const renderItemCard = () =>
  render(
    <ReduxProvider store={mockStore()}>
      <ItemCard item={item} />
    </ReduxProvider>
  );
describe("Tests for <ItemCard/>", () => {
  afterEach(cleanup);

  it("renders without crashing", () => {
    renderItemCard();
  });
  test("ItemCard: Snapshot", () => {
    const component = renderer.create(
      <ReduxProvider store={mockStore()}>
        <ItemCard item={item} />
      </ReduxProvider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('Displays item data"', () => {
    const { getByText } = renderItemCard();
    expect(getByText("Test product title"));
    expect(getByText("â‚¬20.22"));
  });
});
