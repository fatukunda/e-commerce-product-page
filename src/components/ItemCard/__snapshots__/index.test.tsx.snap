// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for <ItemCard/> ItemCard: Snapshot 1`] = `
<div
  className="w-48 bg-white rounded-lg shadow-md dark:border-gray-700"
>
  <img
    alt="Test product title"
    className="p-8 rounded-t-lg"
    src="https://test-image"
  />
  <div
    className="px-5 pb-5"
  >
    <h5
      className="text-xs text-tertiary-light"
    >
      Test product title
    </h5>
    <div
      className="flex justify-between items-center"
    >
      <span
        className="text-l pt-2 font-bold text-gray-900"
      >
        â‚¬
        20.22
      </span>
    </div>
  </div>
</div>
`;
